[FSDP] Add an arg for FSDP __init__

foreveronehundred opened the pull request:
Add an arg, disable_reshard_on_root, for FSDP __init__ to handle the following issue
https://github.com/facebookresearch/fairscale/issues/878
For some cases (models wrapped by autowrap), the parameters (of root modules) needs to be sharded, and reshard_after_forward should not be set to False.
"disable_reshard_on_root" is for users to choose whether to force reshard_after_forward of root modules to be False or not.

## What does this PR do?
Fixes [# 878](https://github.com/facebookresearch/fairscale/issues/878).

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [ ] Did you write any new necessary tests?
  - [x] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A


## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

min-xu-ai commented on fairscale/nn/data_parallel/fully_sharded_data_parallel.py:
```diff
@@ -204,6 +204,11 @@ class FullyShardedDataParallel(nn.Module):
+            if ``True``, reshard_after_forward will be set to False if the module is a root.
+            For some cases, the full parameters of root do not need to be sharded in order to
+            improve performance. It is because the parameters are needed immediately
+            for the backward pass.
```
Can you add a sentence on why do we want to disable this in some cases?

foreveronehundred replied:
I have updated my description to further explain the reason. 
Please tell me if there is still other problem. Thanks.
By the way, the default value **disable_reshard_on_root** is True so that my update will not change the original behavior of reshard_after_forward

min-xu-ai reviewed (approved):
LGTM with a minor comment. Thanks for the PR!

min-xu-ai commented:
@foreveronehundred. Thanks for the PR. Can you sign the CLA. It is just a standard agreement for contribution. Thanks!

foreveronehundred commented:
> 

Done!

foreveronehundred commented:
> @foreveronehundred. Thanks for the PR. Can you sign the CLA. It is just a standard agreement for contribution. Thanks!

Done!

min-xu-ai commented on fairscale/nn/data_parallel/fully_sharded_data_parallel.py:
```diff
@@ -204,6 +204,12 @@ class FullyShardedDataParallel(nn.Module):
             memory but slows training. This is only relevant when resharding
             individual layers.
+        disable_reshard_on_root (bool, Optional):
+            if ``True``, reshard_after_forward will be set to False if the module is a FSDP root module.
+            It helps to improve the performance by setting disable_reshard_on_root to True if a FSDP root
+            module is the root of the entire modules.
+            For that case, we do not reshard full parameters of FSDP modules since those parameters are needed
+            immediately for the backward pass.
```
```suggestion
            If ``True``, ``reshard_after_forward`` will be set to ``False`` if the module is a
            FSDP root module to improve performance.
            If ``False``, the performance will be lower, but it is needed because ....
            Default: True.
```

min-xu-ai replied:
Thanks for the update. I edited it. Can you fill in the reason after "because ...."?

foreveronehundred commented:
Hi @min-xu-ai, may I ask you some questions?
1. I saw that three tests failed. Should I fixed them? How?
2. What should I do next to finish pull request?

min-xu-ai commented:
>     1. I saw that three tests failed. Should I fixed them? How?

Thanks @foreveronehundred. No, the CI errors are due to some other things that we are still trying to figure it out. I can merge this once it is ready.
 
>     2. What should I do next to finish pull request?

To make it ready, can you address this question I raised?
![image](https://user-images.githubusercontent.com/24926999/152650509-17071015-0d3a-484e-b1a2-f96c3dba2e6d.png)

foreveronehundred commented:
> > ```
> > 1. I saw that three tests failed. Should I fixed them? How?
> > ```
> 
> Thanks @foreveronehundred. No, the CI errors are due to some other things that we are still trying to figure it out. I can merge this once it is ready.
> 
> > ```
> > 2. What should I do next to finish pull request?
> > ```
> 
> To make it ready, can you address this question I raised? ![image](https://user-images.githubusercontent.com/24926999/152650509-17071015-0d3a-484e-b1a2-f96c3dba2e6d.png)

Thanks for your advice. I have modified the description with you suggestion. Please tell me if there is any problem.

min-xu-ai commented:
That's great! Thanks for updating! There seems to be a "make doc" error in the cpu test. Can you fix it? Thanks!

foreveronehundred commented:
> That's great! Thanks for updating! There seems to be a "make doc" error in the cpu test. Can you fix it? Thanks!

Hi, could you tell me what is the root cause of the make doc error in the cpu test and how to solve it? 
My update is only for comments so it is very strange that the changes resulted new errors.

min-xu-ai commented:
It is strange. Make doc does parse the comment of the code and generate docs. I will try to fork you change and run it locally myself to see what might be the error. But I perhaps won't get to this in a while.

anj-s commented:
> It is strange. Make doc does parse the comment of the code and generate docs. I will try to fork you change and run it locally myself to see what might be the error. But I perhaps won't get to this in a while.

@min-xu-ai @foreveronehundred  I am seeing similar errors in my PR. I can take a look at this issue and update the PR once I know what is going on.

anj-s commented:
> > It is strange. Make doc does parse the comment of the code and generate docs. I will try to fork you change and run it locally myself to see what might be the error. But I perhaps won't get to this in a while.
> 
> @min-xu-ai @foreveronehundred I am seeing similar errors in my PR. I can take a look at this issue and update the PR once I know what is going on.

Figured it out! We need to update the intersphinx mapping which is currently pointing to something stale. Let me send a PR to fix this and then we can merge your PR after a green run.

anj-s commented:
> > > It is strange. Make doc does parse the comment of the code and generate docs. I will try to fork you change and run it locally myself to see what might be the error. But I perhaps won't get to this in a while.
> > 
> > 
> > @min-xu-ai @foreveronehundred I am seeing similar errors in my PR. I can take a look at this issue and update the PR once I know what is going on.
> 
> Figured it out! We need to update the intersphinx mapping which is currently pointing to something stale. Let me send a PR to fix this and then we can merge your PR after a green run.

Here is the [PR](https://github.com/facebookresearch/fairscale/pull/929) if you want to give it a whirl.
